test/integration/job/job_test.go

```diff
@@ -1026,6 +1029,202 @@ func TestIndexedJob(t *testing.T) {
+
+			// Wait for pods to start up.
+			err = wait.PollImmediate(5*time.Millisecond, wait.ForeverTestTimeout, func() (done bool, err error) {
+				job, err := jobClient.Get(ctx, jobObj.Name, metav1.GetOptions{})
+				if err != nil {
+					return false, err
+				}
+				if job.Status.Active == int32(initialCompletions) {
+					return true, nil
+				}
+				return false, nil
+			})
+			if err != nil {
+				t.Fatalf("Error waiting for Job pods to become active: %v", err)
+			}
+
+			currentCompletions := initialCompletions
+
+			for _, update := range tc.jobUpdates {
+				// Update Job spec if necessary.
+				if update.completions != NoCompletionsUpdate {
+					if jobObj, err = updateJob(ctx, jobClient, jobObj.Name, func(j *batchv1.Job) {
+						j.Spec.Completions = pointer.Int32Ptr(int32(update.completions))
+						j.Spec.Parallelism = pointer.Int32Ptr(int32(update.completions))
+					}); err != nil {
+						// Error expected if feature gate is off
+						if !tc.featureGate {
+							return
+						}
+						t.Fatalf("Failed to update Job: %v", err)
+					}
+					currentCompletions = update.completions
+				}
+
+				// Succeed specified indexes.
+				succeededSet := sets.NewInt()
+				if update.succeededIndexes != nil {
+					for _, idx := range update.succeededIndexes {
+						if err := setJobPhaseForIndex(ctx, clientSet, jobObj, v1.PodSucceeded, idx); err != nil {
+							t.Fatalf("Failed trying to succeed pod with index %d: %v", idx, err)
+						}
+						succeededSet.Insert(idx)
+					}
+					newActive := currentCompletions - len(update.succeededIndexes)
+
+					// initialCompletions == initial parallelism, and active must be <= parallelism.
+					if newActive > currentCompletions && currentCompletions != NoCompletionsUpdate {
+						newActive = currentCompletions
+					}
+					validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+						Active:    newActive,
+						Succeeded: len(update.succeededIndexes),
+						Ready:     pointer.Int32(0),
+					})
+					validateIndexedJobPods(ctx, t, clientSet, jobObj, remainingIndexSet(currentCompletions, succeededSet), update.wantSucceededIndexes)
+				}
+
+				// Fail specified indexes.
+				if update.failedIndexes != nil {
+					for _, idx := range update.failedIndexes {
+						if err := setJobPhaseForIndex(ctx, clientSet, jobObj, v1.PodFailed, idx); err != nil {
+							t.Fatalf("Failed trying to fail pod with index %d: %v", idx, err)
+						}
+					}
+					remainingIndexes := remainingIndexSet(currentCompletions, succeededSet)
+					newActive := len(remainingIndexes)
+
+					// initialCompletions == initial parallelism, and active must be <= parallelism.
+					if newActive > currentCompletions && currentCompletions != NoCompletionsUpdate {
+						newActive = currentCompletions
+					}
+					validateJobPodsStatus(ctx, t, clientSet, jobObj, podsByStatus{
+						Active:    newActive,
+						Succeeded: len(succeededSet),
+						Failed:    update.wantFailed,
+						Ready:     pointer.Int32(0),
+					})
```

Comment:
ditto, we should verify wantFailed all the time on each step, not only when failedIndexes is set.

Reply:
I see, that makes sense now. Done.